### PR TITLE
Shield memberships blocks for unconnected jetpack sites

### LINKED ISSUE #39308: Blocks: Membership Endpoint Request WordPress.com Connection in Editor
### Impacted plugin

Jetpack

### What

Prevent WordPress.com Notice in Editor when no Jetpack connection is present

### How

When an editor with `edit_posts` permissions is logged in but has not connected Jetpack visiting Post > Add New results in a message to connect Jetpack:
Screenshot 2024-09-09 at 15 53 33

This appears to stem from the `membership` endpoint:
`https://example.com/wp-json/wpcom/v2/memberships/status?source=gutenberg&type=all&is_editable=true&_locale=user`

My browser returns a 500 error and the following response body:

```
{"code":"missing_token","message":"Please connect your user account to WordPress.com","data":404}
```

Source: https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/memberships.php#L243-L248

Most Jetpack blocks only ask you to connect Jetpack when they are inserted:
Screenshot 2024-09-09 at 16 28 53

Alerting the editor to connect Jetpack on every page load seems a bit over zealous and if a reason is not provided as to why, the alert seems unhelpful. 

## Replicate:

1) Create a Jetpack site with a primary Jetpack connection
2) Log in as an editor who is not connected to Jetpack
3) Visit Post > Add New 
4) See the Connect Jetpack Message

## Current Workaround
Disable Jetpack Blocks

### PR DESCRIPTION
**Pull request by millerf**
Fixes https://github.com/Automattic/jetpack/issues/39308

## Proposed changes:

* Do not register Memberships blocks that actively use Jetpack connection when the site is not connected to Jetpack
* [front-end] Use of utility function `Jetpack_Memberships::is_enabled_jetpack_recurring_payments()`
* [admin] Check for proper JP connection

Note: Simple payments blocks. (aka as Paypal button) is not shielded as it does not need a JP connection to be working

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

See https://github.com/Automattic/jetpack/issues/39308#issuecomment-2340110903

## Does this pull request change what data or activity we track or use?

No

## Testing instructions:

* Create a JP site with a connection
  * Make sure Donations, Payment Button and Premium content buttons are now available
  * Create a post with a paid-content, a subscription and a simple-payments block 
  * Make sure the  blocks is available on the front-end
* Disconnect the site from `wp-admin/admin.php?page=my-jetpack`
  * Disconnect the site
  * Make sure the block is not available in the admin
  * Make sure the blocks are still available on the previously-created post

### CODE REVIEW

File: projects/plugins/jetpack/modules/memberships/class-jetpack-memberships.php
Comment (pottedmeat) [inline projects/plugins/jetpack/modules/memberships/class-jetpack-memberships.php:759]: I think this is reversed and it should return true when `is_admin()` is true.

Comment (pottedmeat) [reply projects/plugins/jetpack/modules/memberships/class-jetpack-memberships.php:759]: Okay, I understand this now. `is_admin` refers to an admin page.

**Review by pottedmeat — APPROVED** [review]
Code looks good and this tested as expected.

**Review by jeherve — APPROVED** [review]
This seems to test well for me on WordPress.com Simple.

File: projects/plugins/jetpack/modules/memberships/class-jetpack-memberships.php
Comment (millerf) [reply projects/plugins/jetpack/modules/memberships/class-jetpack-memberships.php:759]: > I think this is reversed and it should return true when is_admin() is true.

You scared me for a bit: it tested fine so I was wondering what I did wrong here :)

**Comment by CodeyGuyDylan** [comment]
@millerf and @jeherve, as part of this project (pbNhbs-bHl-p2) we are trying to make the connection nudge consistent across Jetpack including the editor with the Jetpack blocks. Read this comment for more context (pbNhbs-bHl-p2#comment-23877) but we were wondering if we could initialize the blocks in the case of a site without a connected user and show a connection banner instead? That way the user still knows they exist and can still find them, but is directed towards what to do to resolve the issue.

cc: @robertsreberski 

The banner would look like it does in [this PR](https://github.com/Automattic/jetpack/pull/39638) and is currently used in the AI block.

**Comment by CodeyGuyDylan** [comment]
FYI, [this PR](https://github.com/Automattic/jetpack/pull/39647) is what I have in mind for this

**Comment by jeherve** [comment]
> we were wondering if we could initialize the blocks in the case of a site without a connected user and show a connection banner instead?

That seems like a better user experience indeed. We would just need to account for cases like this PR solved: some blocks may need the user connection to display their original state or to make an API request that will return information the block needs. For such blocks, it would be better if there was some kind of placeholder state imo.

**Comment by CodeyGuyDylan** [comment]
> That seems like a better user experience indeed. We would just need to account for cases like this PR solved: some blocks may need the user connection to display their original state or to make an API request that will return information the block needs. For such blocks, it would be better if there was some kind of placeholder state imo.

In the PR that I shared, the connection banner component there is a prop called "explanation" that displays a message as to what they'd be connecting their user account for (ex. "Connect your WordPress.com account to enable payment buttons on your site.")

Was that the kind of thing you had in mind or am I misunderstanding?

**Comment by CodeyGuyDylan** [comment]
The PR I opened got approved but I'll wait for confirmation that is what you meant before merging 😄

**Pull request merged**
